### DISCUSSION TITLE
Is it possible to configure the Aedes broker to allow client connections over both TCP and WS?

### DISCUSSION THREAD

**Discussion by ninjaclown123** [moscajs/aedes#938] [Q&A] [answered]
I am looking to configure my Aedes broker to achieve the scenario in the title. I am currently in a scenario where I have a client that can only connect to the broker over websockets and another client that connects over HTTP. 

It might be possible for me to configure the second client to connect over websockets but I would like to explore what I mentioned above before proceeding to configure the client to connect over HTTP.

I am currently able to achieve the intended behavior I want but on two different ports i.e, the server is listening for HTTP connections on port 1883 and for websocket connections on port 8888. I have attached the current implementation below:

![snippet](https://github.com/moscajs/aedes/assets/135505235/ad088e22-030f-4fb9-8b10-5dc3f2adf068)

**Comment by robertsLando** [comment]
I usually bind the ws/wss to the express server and then create a separete broker connection for plain tcp/tls. This is how I bind the ws server:

```ts
import { createWebSocketStream, Server as WsServer } from 'ws'

const wsServer = new WsServer({ server: // your http server })

wsServer.on('error', err => {
	console.error(err, 'WebSocket error')
})

wsServer.on('connection', (conn, req) => {
	const stream = createWebSocketStream(conn)
	stream._socket = conn._socket
	broker.handle(stream, req)
})
```

**Reply by ninjaclown123** [reply]
Thanks for the response, but I'm not sure about your answer. 

I currently have the aedes broker working alongside express already like in the screenshot. I'm able to make REST requests to my server and I'm also able to connect MQTT clients over websockets. I can also connect clients over HTTP but through another port. I want to be able to allow clients to connect over WS and TCP through the same port, is this possible? 

What I gather from your answer is that after I create my express server, I can bind the ws server to the express server using your example. And then I'm not quite sure how I can allow clients to connect over TCP by "creating a separate broker connection". Will this be possible on the same port as the one used for websockets?

**Reply by robertsLando** [reply, marked as answer]
Websockets can only be enabled on http servers not tcp streams so you cannot for use the server you create to listen to 1883 to also accept websocket at least for what I know
